os/fs/driver/mtd/smart.c

```diff
@@ -3489,23 +3564,73 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
 #ifdef CONFIG_SMARTFS_BAD_SECTOR
 			if (dev->badSectorList[x] == FALSE) {
 #endif
-				physicalsector = x;
-				dev->lastallocblock = allocblock;
-				break;
+				ret = MTD_READ(dev->mtd, readaddr, dev->mtdBlksPerSector * dev->geo.blocksize, sector_buff);
+				if (ret != dev->mtdBlksPerSector * dev->geo.blocksize) {
+					fdbg("Error reading physical sector %d\n",physicalsector);
+					kmm_free(sector_buff);
+					return -1;
+				}
+				for (i = 0; i < dev->mtdBlksPerSector * dev->geo.blocksize; i++) {
+					if (sector_buff[i] != 0xff) {
+						break;
+					}
+				}
+
+				if (i == dev->mtdBlksPerSector * dev->geo.blocksize) {
+					physicalsector = x;
+					dev->lastallocblock = allocblock;
+					break;
+				} else {
+					bitflipped = TRUE;
+					fdbg("bit flip occur %d offset %d byte %x\n", x, i, sector_buff[i]);
+					fdbg("set Released and commited to Erase%d\n", x);
+#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
+					header.status = header.status & ~(SMART_STATUS_COMMITTED | SMART_STATUS_RELEASED);
+#else
+					header.status = header.status | SMART_STATUS_COMMITTED | SMART_STATUS_RELEASED;
+#endif
+					ret = smart_bytewrite(dev, readaddr + offsetof(struct smart_sect_header_s, status), 1, &header.status);
+					if (ret < 0) {
+						fdbg("Error %d releasing corrupted sector\n", -ret);
+						goto error;
+					}
+#ifdef CONFIG_MTD_SMART_PACK_COUNTS
+					smart_add_count(dev, dev->freecount, x / dev->sectorsPerBlk, -1);
+					smart_add_count(dev, dev->releasecount, allocblock, 1);
+#else
+					dev->freecount[x / dev->sectorsPerBlk]--;
+					dev->releasecount[allocblock]++;
+#endif
+					dev->freesectors--;
+					dev->releasesectors++;
+				}
 #ifdef CONFIG_SMARTFS_BAD_SECTOR
 			}
 #endif
 		}
 	}
 
+error:
 	if (physicalsector == 0xFFFF) {
-		dbg("Program bug!  Expected a free sector\n");
+		if (bitflipped) {
+			fdbg("retry allocation \n");
+			canrelocate = FALSE;
+			goto retry;
```

Comment:
Now I can see bug in my code.
If we move to retry here or below, there's leakage because we didn't free sector_buff.
Checking condition or freeing code should be added...

Reply:
Yes, It is added in https://github.com/Samsung/TizenRT/pull/524/commits/592e0aaea6a6fbbc99347cd3cc7ffce7b044bffd